File: rust/src/browser/browser_cookie_credentials.rs
//! OS credential-store access for installed Chromium cookie import.

use std::fs;
use std::path::Path;
use std::process::Command;

use anyhow::{anyhow, Context, Result};
use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use serde_json::Value;

struct SafeStorageIdentity {
    application: &'static str,
    folder: &'static str,
    service: &'static str,
}

fn safe_storage_identity(browser: &str) -> Result<SafeStorageIdentity> {
    let identity = match browser {
        "brave" => SafeStorageIdentity {
            application: "brave",
            folder: "Brave Keys",
            service: "Brave Safe Storage",
        },
        "chrome" => SafeStorageIdentity {
            application: "chrome",
            folder: "Chrome Keys",
            service: "Chrome Safe Storage",
        },
        "chromium" => SafeStorageIdentity {
            application: "chromium",
            folder: "Chromium Keys",
            service: "Chromium Safe Storage",
        },
        "edge" => SafeStorageIdentity {
            application: "microsoft-edge",
            folder: "Microsoft Edge Keys",
            service: "Microsoft Edge Safe Storage",
        },
        _ => return Err(anyhow!("No Safe Storage identity is known for {browser}")),
    };
    Ok(identity)
}

fn run_credential_command(command: &str, arguments: &[&str]) -> Result<String> {
    let output = Command::new(command)
        .args(arguments)
        .output()
        .with_context(|| format!("Could not start {command}"))?;
    if !output.status.success() {
        return Err(anyhow!("{command} exited with {}", output.status));
    }
    String::from_utf8(output.stdout)
        .context("credential command returned invalid UTF-8")
        .map(|value| value.trim().to_string())
}

pub(crate) fn read_safe_storage_password(browser: &str, platform: &str) -> Result<String> {
    let identity = safe_storage_identity(browser)?;
    if platform == "darwin" {
        let password = run_credential_command(
            "security",
            &["find-generic-password", "-w", "-s", identity.service],
        )?;
        return (!password.is_empty())
            .then_some(password)
            .ok_or_else(|| anyhow!("{} returned an empty password", identity.service));
    }
    if platform == "linux" {
        if let Ok(password) = run_credential_command(
            "secret-tool",
            &["lookup", "application", identity.application],
        ) {
            if !password.is_empty() {
                return Ok(password);
            }
        }
        if let Ok(password) = run_credential_command(
            "kwallet-query",
            &["-r", identity.service, "-f", identity.folder, "kdewallet"],
        ) {
            if !password.is_empty() {
                return Ok(password);
            }
        }
        return Err(anyhow!(
            "Could not read {} from libsecret or KWallet; install secret-tool or unlock the browser key store",
            identity.service
        ));
    }
    Err(anyhow!("Safe Storage passwords are not used on {platform}"))
}

const DPAPI_SCRIPT: &str = concat!(
    "$inputBytes=[Convert]::FromBase64String($args[0]);",
    "$outputBytes=[Security.Cryptography.ProtectedData]::Unprotect(",
    "$inputBytes,$null,[Security.Cryptography.DataProtectionScope]::CurrentUser);",
    "[Convert]::ToBase64String($outputBytes)"
);

pub(crate) fn decrypt_windows_dpapi(encrypted: &[u8]) -> Result<Vec<u8>> {
    let encoded = BASE64.encode(encrypted);
    let output = run_credential_command(
        "powershell.exe",
        &[
            "-NoLogo",
            "-NoProfile",
            "-NonInteractive",
            "-Command",
            DPAPI_SCRIPT,
            &encoded,
        ],
    )?;
    BASE64
        .decode(output)
        .context("DPAPI command returned invalid base64")
}

pub(crate) fn read_windows_encryption_key(local_state_path: &Path) -> Result<Vec<u8>> {
    let contents = fs::read_to_string(local_state_path).with_context(|| {
        format!(
            "Could not read Chromium Local State {}",
            local_state_path.display()
        )
    })?;
    let state: Value = serde_json::from_str(&contents).context("Invalid Chromium Local State")?;
    let encoded = state
        .pointer("/os_crypt/encrypted_key")
        .and_then(Value::as_str)
        .ok_or_else(|| anyhow!("Chromium Local State has no os_crypt.encrypted_key"))?;
    let encrypted_key = BASE64
        .decode(encoded)
        .context("Chromium Local State encrypted_key is not base64")?;
    let protected = encrypted_key
        .strip_prefix(b"DPAPI")
        .ok_or_else(|| anyhow!("Chromium Local State key does not have a DPAPI prefix"))?;
    decrypt_windows_dpapi(protected)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kwallet_identity_uses_chromium_product_casing() {
        let identity = safe_storage_identity("chrome").unwrap();
        assert_eq!(identity.folder, "Chrome Keys");
        assert_eq!(identity.service, "Chrome Safe Storage");
    }
}

File: js/src/browser/browser-cookie-credentials.js
import { execFile as execFileCallback } from 'node:child_process';
import { readFile } from 'node:fs/promises';
import { promisify } from 'node:util';

const execFile = promisify(execFileCallback);

const SAFE_STORAGE = {
  brave: {
    application: 'brave',
    folder: 'Brave Keys',
    service: 'Brave Safe Storage',
  },
  chrome: {
    application: 'chrome',
    folder: 'Chrome Keys',
    service: 'Chrome Safe Storage',
  },
  chromium: {
    application: 'chromium',
    folder: 'Chromium Keys',
    service: 'Chromium Safe Storage',
  },
  edge: {
    application: 'microsoft-edge',
    folder: 'Microsoft Edge Keys',
    service: 'Microsoft Edge Safe Storage',
  },
};

async function runCredentialCommand(command, args, environment) {
  const { stdout } = await execFile(command, args, {
    encoding: 'utf8',
    env: environment,
    maxBuffer: 1024 * 1024,
    windowsHide: true,
  });
  return stdout.trim();
}

async function readLinuxSafeStoragePassword(browser, environment, runCommand) {
  const identity = SAFE_STORAGE[browser];
  try {
    const password = await runCommand(
      'secret-tool',
      ['lookup', 'application', identity.application],
      environment
    );
    if (password) {
      return password;
    }
  } catch {
    // Try KWallet before reporting that v11 key storage is unavailable.
  }
  try {
    const password = await runCommand(
      'kwallet-query',
      ['-r', identity.service, '-f', identity.folder, 'kdewallet'],
      environment
    );
    if (password) {
      return password;
    }
  } catch {
    // Fall through to the actionable error below.
  }
  throw new Error(
    `Could not read ${identity.service} from libsecret or KWallet; install secret-tool or unlock the browser key store`
  );
}

/** Read a Chromium Safe Storage password from the platform credential store. */
export async function readSafeStoragePassword({
  browser,
  platform = process.platform,
  environment = process.env,
  runCredentialCommand: runCommand = runCredentialCommand,
}) {
  const identity = SAFE_STORAGE[browser];
  if (!identity) {
    throw new Error(`No Safe Storage identity is known for ${browser}`);
  }
  if (platform === 'darwin') {
    const password = await runCommand(
      'security',
      ['find-generic-password', '-w', '-s', identity.service],
      environment
    );
    if (!password) {
      throw new Error(`${identity.service} returned an empty password`);
    }
    return password;
  }
  if (platform === 'linux') {
    return readLinuxSafeStoragePassword(browser, environment, runCommand);
  }
  throw new Error(`Safe Storage passwords are not used on ${platform}`);
}

const DPAPI_SCRIPT = [
  '$inputBytes=[Convert]::FromBase64String($args[0])',
  '$outputBytes=[Security.Cryptography.ProtectedData]::Unprotect(',
  '  $inputBytes,$null,',
  '  [Security.Cryptography.DataProtectionScope]::CurrentUser)',
  '[Convert]::ToBase64String($outputBytes)',
].join(';');

/** Decrypt bytes with Windows DPAPI in the current user's security context. */
export async function decryptWindowsDpapi(
  encryptedValue,
  { environment = process.env } = {}
) {
  const output = await runCredentialCommand(
    'powershell.exe',
    [
      '-NoLogo',
      '-NoProfile',
      '-NonInteractive',
      '-Command',
      DPAPI_SCRIPT,
      Buffer.from(encryptedValue).toString('base64'),
    ],
    environment
  );
  return Buffer.from(output, 'base64');
}

/** Recover the legacy AES-GCM key from a Chromium Local State file. */
export async function readWindowsEncryptionKey({
  localStatePath,
  environment = process.env,
  decryptDpapi = decryptWindowsDpapi,
}) {
  let state;
  try {
    state = JSON.parse(await readFile(localStatePath, 'utf8'));
  } catch (error) {
    throw new Error(`Could not read Chromium Local State: ${error.message}`, {
      cause: error,
    });
  }
  const encoded = state.os_crypt?.encrypted_key;
  if (!encoded) {
    throw new Error('Chromium Local State has no os_crypt.encrypted_key');
  }
  const encryptedKey = Buffer.from(encoded, 'base64');
  if (encryptedKey.subarray(0, 5).toString('ascii') !== 'DPAPI') {
    throw new Error('Chromium Local State key does not have a DPAPI prefix');
  }
  return decryptDpapi(encryptedKey.subarray(5), { environment });
}
